databroker/broker.py

```diff
@@ -880,3 +902,178 @@ def _munge_time(t, timezone):
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+
+def store_dec(db, external_writers=None):
+    """Decorate a generator of documents to save them to the databases.
+
+    The input stream of (name, document) pairs passes through unchanged.
+    As a side effect, documents are inserted into the databases and external
+    files may be written.
+
+    Parameters
+    ----------
+    db: ``databroker.Broker`` instance
+        The databroker to store the documents in, must have writeable
+        metadatastore and writeable filestore if ``external`` is not empty.
+    external_writers : dict
+        Maps data keys to a ``WriterClass``, which is responsible for writing
+        data to disk and creating a record in filestore. It will be
+        instantiated (possible multiple times) with the argument ``db.fs``.
+        If it requires additional arguments, use ``functools.partial`` to
+        produce a callable that requires only ``db.fs`` to instantiate the
+        ``WriterClass``.
+        """
+    if external_writers is None:
+        external_writers = {}  # {'name': WriterClass}
+
+    def wrap(f):
+        def wrapped_f(*args, **kwargs):
```

Comment:
This looks like it assumes you will never get mixed event streams, that will not be true for 'live' data.